tests/flow/test_hnswlib.py

```diff
@@ -518,11 +450,481 @@ def test_multi_range_query():
+        assert set(vectors_s.flatten()) == set(vectors_p.flatten())
+
+    print(f"Inserting {num_elements} vectors of dim {dim} into multi-HNSW in parallel ({per_label} vectors per label)"
+          f" took {parallel_insert_time} seconds")
+    print(f"Got {sequential_insert_time/parallel_insert_time} times improvement using {n_threads} threads\n")
+    assert sequential_insert_time/parallel_insert_time > expected_speedup
+
+    # Run queries over the multi-index
+    query_data = np.random.random((num_queries, dim))
+
+    # Sequential search as the baseline
+    total_search_time = 0
+    total_correct = 0
+    total_res_bf = []  # save the ground truth res_labels for every query
+    total_res_hnsw_sequential = []  # save sequential search results, tuple of (res_labels, res_dists) for every query
+    for i, query in enumerate(query_data):
+        start = time.time()
+        res_labels, res_dists = multi_index.knn_query(query, k)
+        total_search_time += time.time() - start
+        res_labels_bf, res_dists_bf = bf_index.knn_query(query, k)
+        total_res_hnsw_sequential.append((res_labels[0], res_dists[0]))
+        total_res_bf.append(res_labels_bf[0])
+        total_correct += len(set(res_labels[0]).intersection(set(res_labels_bf[0])))
+    print(f"Running sequential search, got {total_correct / (k * num_queries)} recall on {num_queries} queries,"
+          f" average query time is {total_search_time / num_queries} seconds")
+
+    start = time.time()
+    res_labels_parallel, res_dists_parallel = parallel_multi_index.knn_parallel(query_data, k, num_threads=n_threads)
+    total_search_time_parallel = time.time() - start
+
+    total_correct_parallel = 0
+    for res_labels, ground_truth in zip(res_labels_parallel, total_res_bf):
+        total_correct_parallel += len(set(res_labels).intersection(set(ground_truth)))
+
+    print(f"Running parallel search, got {total_correct_parallel / (k * num_queries)} recall on {num_queries} queries,"
+          f" average query time is {total_search_time_parallel / num_queries} seconds")
+    print(f"Got {total_search_time / total_search_time_parallel} times improvement in runtime using"
+          f" {n_threads} threads\n")
+    assert total_correct_parallel >= total_correct*0.95
+    assert total_search_time/total_search_time_parallel >= expected_speedup
+
+    # Insert vectors to the index and search in parallel.
+    parallel_multi_index = create_hnsw_index(dim, num_elements, metric, data_type, m=32, ef_runtime=200,
+                                             is_multi=True)
+    assert parallel_multi_index.index_size() == 0
+
+    def insert_vectors():
+        parallel_multi_index.add_vector_parallel(data, labels, num_threads=int(n_threads/2))
+
+    res_labels_g = np.zeros((num_queries, dim))
+
+    def run_queries():
+        nonlocal res_labels_g
+        res_labels_g, _ = parallel_multi_index.knn_parallel(query_data, k, num_threads=int(n_threads/2))
```

Comment:
why not to use searchKnnParallel?

Reply:
what do you mean? `searchKnnParallel` is an internal function in the Python bindings. `knn_parallel` is the method name that we expose in the Python module...